{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [],
   "source": [
    "#! pip install scanpy anndata h5py scvi-tools umap-learn\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/albert.baichorov/.venv/lib/python3.10/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html\n",
      "  from .autonotebook import tqdm as notebook_tqdm\n",
      "/home/albert.baichorov/.venv/lib/python3.10/site-packages/docrep/decorators.py:43: SyntaxWarning: 'param_categorical_covariate_keys' is not a valid key!\n",
      "  doc = func(self, args[0].__doc__, *args[1:], **kwargs)\n",
      "/home/albert.baichorov/.venv/lib/python3.10/site-packages/docrep/decorators.py:43: SyntaxWarning: 'param_continuous_covariate_keys' is not a valid key!\n",
      "  doc = func(self, args[0].__doc__, *args[1:], **kwargs)\n"
     ]
    }
   ],
   "source": [
    "import anndata as ad\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import scvi\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "AnnData object with n_obs × n_vars = 538266 × 22091\n",
      "    obs: 'orig.ident', 'nCount_RNA', 'nFeature_RNA', 'percent.mt', 'percent.HB', 'percent.RPS', 'S.Score', 'G2M.Score', 'Phase', 'umis_TCR', 'umis_BCR', 'primary_type', 'secondary_type', 'sequential_group', 'eight_group', 'sampleName', 'age'\n",
      "    uns: 'version'\n",
      "    layers: 'RNA_counts', 'RNA_data'\n"
     ]
    }
   ],
   "source": [
    "\n",
    "adata = ad.read_h5ad(\"scRNA-seqProcessedLabelledObject.h5ad\")\n",
    "print(adata)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AL627309.1</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL669831.5</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>LINC00115</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AL645608.3</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NOC2L</th>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "Empty DataFrame\n",
       "Columns: []\n",
       "Index: [AL627309.1, AL669831.5, LINC00115, AL645608.3, NOC2L]"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.var.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>orig.ident</th>\n",
       "      <th>nCount_RNA</th>\n",
       "      <th>nFeature_RNA</th>\n",
       "      <th>percent.mt</th>\n",
       "      <th>percent.HB</th>\n",
       "      <th>percent.RPS</th>\n",
       "      <th>S.Score</th>\n",
       "      <th>G2M.Score</th>\n",
       "      <th>Phase</th>\n",
       "      <th>umis_TCR</th>\n",
       "      <th>umis_BCR</th>\n",
       "      <th>primary_type</th>\n",
       "      <th>secondary_type</th>\n",
       "      <th>sequential_group</th>\n",
       "      <th>eight_group</th>\n",
       "      <th>sampleName</th>\n",
       "      <th>age</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGCGTACA-1-sample42</th>\n",
       "      <td>seurat_object</td>\n",
       "      <td>7685.0</td>\n",
       "      <td>1889.0</td>\n",
       "      <td>2.446324</td>\n",
       "      <td>0.0</td>\n",
       "      <td>47.690306</td>\n",
       "      <td>0.008262</td>\n",
       "      <td>0.003594</td>\n",
       "      <td>S</td>\n",
       "      <td>26</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>CD4T</td>\n",
       "      <td>CD4_TEM_ANXA1</td>\n",
       "      <td>60y</td>\n",
       "      <td>50-60y</td>\n",
       "      <td>sample42</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGGAGTAA-1-sample42</th>\n",
       "      <td>seurat_object</td>\n",
       "      <td>6206.0</td>\n",
       "      <td>1911.0</td>\n",
       "      <td>3.045440</td>\n",
       "      <td>0.0</td>\n",
       "      <td>36.545279</td>\n",
       "      <td>0.008496</td>\n",
       "      <td>-0.042660</td>\n",
       "      <td>S</td>\n",
       "      <td>11</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>CD4T</td>\n",
       "      <td>CD4_TEM_ANXA1</td>\n",
       "      <td>60y</td>\n",
       "      <td>50-60y</td>\n",
       "      <td>sample42</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAGCGATCC-1-sample42</th>\n",
       "      <td>seurat_object</td>\n",
       "      <td>4726.0</td>\n",
       "      <td>1764.0</td>\n",
       "      <td>4.485823</td>\n",
       "      <td>0.0</td>\n",
       "      <td>26.195514</td>\n",
       "      <td>0.089389</td>\n",
       "      <td>0.009794</td>\n",
       "      <td>S</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>CD8T</td>\n",
       "      <td>CD8_TEM_GNLY</td>\n",
       "      <td>60y</td>\n",
       "      <td>50-60y</td>\n",
       "      <td>sample42</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCATATGGTC-1-sample42</th>\n",
       "      <td>seurat_object</td>\n",
       "      <td>7870.0</td>\n",
       "      <td>2112.0</td>\n",
       "      <td>3.481576</td>\n",
       "      <td>0.0</td>\n",
       "      <td>42.757306</td>\n",
       "      <td>-0.002148</td>\n",
       "      <td>0.009421</td>\n",
       "      <td>G2M</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>35</td>\n",
       "      <td>B</td>\n",
       "      <td>B_Memory</td>\n",
       "      <td>60y</td>\n",
       "      <td>50-60y</td>\n",
       "      <td>sample42</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCATTGAGCT-1-sample42</th>\n",
       "      <td>seurat_object</td>\n",
       "      <td>9081.0</td>\n",
       "      <td>2095.0</td>\n",
       "      <td>5.241713</td>\n",
       "      <td>0.0</td>\n",
       "      <td>46.558749</td>\n",
       "      <td>0.000482</td>\n",
       "      <td>0.006435</td>\n",
       "      <td>G2M</td>\n",
       "      <td>-2147483648</td>\n",
       "      <td>14</td>\n",
       "      <td>B</td>\n",
       "      <td>B_Memory</td>\n",
       "      <td>60y</td>\n",
       "      <td>50-60y</td>\n",
       "      <td>sample42</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                orig.ident  nCount_RNA  nFeature_RNA  \\\n",
       "AAACCTGAGGCGTACA-1-sample42  seurat_object      7685.0        1889.0   \n",
       "AAACCTGAGGGAGTAA-1-sample42  seurat_object      6206.0        1911.0   \n",
       "AAACCTGCAGCGATCC-1-sample42  seurat_object      4726.0        1764.0   \n",
       "AAACCTGCATATGGTC-1-sample42  seurat_object      7870.0        2112.0   \n",
       "AAACCTGCATTGAGCT-1-sample42  seurat_object      9081.0        2095.0   \n",
       "\n",
       "                             percent.mt  percent.HB  percent.RPS   S.Score  \\\n",
       "AAACCTGAGGCGTACA-1-sample42    2.446324         0.0    47.690306  0.008262   \n",
       "AAACCTGAGGGAGTAA-1-sample42    3.045440         0.0    36.545279  0.008496   \n",
       "AAACCTGCAGCGATCC-1-sample42    4.485823         0.0    26.195514  0.089389   \n",
       "AAACCTGCATATGGTC-1-sample42    3.481576         0.0    42.757306 -0.002148   \n",
       "AAACCTGCATTGAGCT-1-sample42    5.241713         0.0    46.558749  0.000482   \n",
       "\n",
       "                             G2M.Score Phase    umis_TCR    umis_BCR  \\\n",
       "AAACCTGAGGCGTACA-1-sample42   0.003594     S          26 -2147483648   \n",
       "AAACCTGAGGGAGTAA-1-sample42  -0.042660     S          11 -2147483648   \n",
       "AAACCTGCAGCGATCC-1-sample42   0.009794     S -2147483648 -2147483648   \n",
       "AAACCTGCATATGGTC-1-sample42   0.009421   G2M -2147483648          35   \n",
       "AAACCTGCATTGAGCT-1-sample42   0.006435   G2M -2147483648          14   \n",
       "\n",
       "                            primary_type secondary_type sequential_group  \\\n",
       "AAACCTGAGGCGTACA-1-sample42         CD4T  CD4_TEM_ANXA1              60y   \n",
       "AAACCTGAGGGAGTAA-1-sample42         CD4T  CD4_TEM_ANXA1              60y   \n",
       "AAACCTGCAGCGATCC-1-sample42         CD8T   CD8_TEM_GNLY              60y   \n",
       "AAACCTGCATATGGTC-1-sample42            B       B_Memory              60y   \n",
       "AAACCTGCATTGAGCT-1-sample42            B       B_Memory              60y   \n",
       "\n",
       "                            eight_group sampleName age  \n",
       "AAACCTGAGGCGTACA-1-sample42      50-60y   sample42  60  \n",
       "AAACCTGAGGGAGTAA-1-sample42      50-60y   sample42  60  \n",
       "AAACCTGCAGCGATCC-1-sample42      50-60y   sample42  60  \n",
       "AAACCTGCATATGGTC-1-sample42      50-60y   sample42  60  \n",
       "AAACCTGCATTGAGCT-1-sample42      50-60y   sample42  60  "
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Описание колонок в аннотированных данных (AnnData.obs)\n",
    "\n",
    "Данный датасет представляет собой результаты одноклеточного РНК-секвенирования (scRNA-seq), содержащие аннотации клеток. Ниже приведены описания колонок:\n",
    "\n",
    "- **`orig.ident`** – Идентификатор исходного образца или объекта, откуда была получена клетка (например, `seurat_object`).\n",
    "- **`nCount_RNA`** – Общее количество молекул РНК (UMI) в клетке. Показывает уровень экспрессии.\n",
    "- **`nFeature_RNA`** – Количество уникальных генов, детектированных в клетке. Отражает транскрипционное разнообразие.\n",
    "- **`percent.mt`** – Доля митохондриальных генов от общего числа РНК. Высокое значение может указывать на стресс или гибель клетки.\n",
    "- **`percent.HB`** – Доля генов, связанных с гемоглобином, от общего количества РНК. Полезно для анализа кровяных клеток.\n",
    "- **`percent.RPS`** – Доля рибосомных белков RPS. Высокие значения могут указывать на активный биосинтез белков.\n",
    "- **`S.Score`** – Оценка активности фазы S клеточного цикла (синтез ДНК). Положительное значение означает активное деление.\n",
    "- **`G2M.Score`** – Оценка активности фазы G2/M клеточного цикла (подготовка к митозу).\n",
    "- **`Phase`** – Определённая фаза клеточного цикла: $G1$, $S$ или $G2M$.\n",
    "- **`umis_TCR`** – Количество UMI, относящихся к генам T-клеточного рецептора (TCR). Используется для идентификации Т-клеток.\n",
    "- **`umis_BCR`** – Количество UMI, относящихся к генам B-клеточного рецептора (BCR). Позволяет выделить B-клетки.\n",
    "- **`primary_type`** – Основной тип клетки (например, CD4+ T-клетки, CD8+ T-клетки, B-клетки).\n",
    "- **`secondary_type`** – Более детальная аннотация клеточного типа, например, `CD4_TEM_ANXA1` (эффекторные CD4+ Т-клетки, экспрессирующие ANXA1) или `B_Memory` (памятные B-клетки).\n",
    "- **`sequential_group`** – Категориальный признак возрастной группы, например, `60y`.\n",
    "- **`eight_group`** – Альтернативная возрастная группировка, например, `50-60y`.\n",
    "- **`sampleName`** – Идентификатор образца (например, `sample42`), указывающий, откуда была получена клетка.\n",
    "- **`age`** – Возраст донора в годах (например, `60`).\n",
    "\n",
    "Данные представлены в формате AnnData и могут использоваться для анализа клеточных популяций, идентификации клеточных состояний и сравнения между разными возрастными группами.\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Column Descriptions in Annotated Data (AnnData.obs)\n",
    "\n",
    "This dataset represents the results of single-cell RNA sequencing (scRNA-seq) with cell annotations. Below is a description of the columns:\n",
    "\n",
    "- **`orig.ident`** – Identifier of the original sample or dataset from which the cell was obtained (e.g., `seurat_object`).\n",
    "- **`nCount_RNA`** – Total count of RNA molecules (UMIs) detected per cell. Indicates the expression level.\n",
    "- **`nFeature_RNA`** – Number of unique genes detected in each cell. Reflects transcriptional diversity.\n",
    "- **`percent.mt`** – Percentage of mitochondrial genes in the total RNA count. High values may indicate stressed or dying cells.\n",
    "- **`percent.HB`** – Percentage of hemoglobin-related genes in the total RNA count. Useful for analyzing blood cells.\n",
    "- **`percent.RPS`** – Percentage of ribosomal protein S (RPS) genes in the total RNA count. High values may indicate increased protein biosynthesis.\n",
    "- **`S.Score`** – Score representing the cell cycle **S-phase** (DNA synthesis phase). Positive values suggest active division.\n",
    "- **`G2M.Score`** – Score representing the **G2/M phase** of the cell cycle (preparation for mitosis).\n",
    "- **`Phase`** – Assigned cell cycle phase: $G1$, $S$, or $G2M$.\n",
    "- **`umis_TCR`** – Number of UMIs mapped to T-cell receptor (TCR) genes. Used to identify T-cell populations.\n",
    "- **`umis_BCR`** – Number of UMIs mapped to B-cell receptor (BCR) genes. Used to identify B-cell populations.\n",
    "- **`primary_type`** – Broad classification of the cell type (e.g., CD4+ T cells, CD8+ T cells, B cells).\n",
    "- **`secondary_type`** – More detailed subtype of the cell, such as `CD4_TEM_ANXA1` (effector memory CD4+ T cells expressing ANXA1) or `B_Memory` (memory B cells).\n",
    "- **`sequential_group`** – Categorical age group label, such as `60y`.\n",
    "- **`eight_group`** – Alternative age grouping format, such as `50-60y`.\n",
    "- **`sampleName`** – Sample or dataset identifier (e.g., `sample42`), indicating the source of the cell.\n",
    "- **`age`** – Age of the donor in years (e.g., `60`).\n",
    "\n",
    "The data is stored in the AnnData format and can be used for analyzing cellular populations, identifying cell states, and comparing different age groups.\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>AL627309.1</th>\n",
       "      <th>AL669831.5</th>\n",
       "      <th>LINC00115</th>\n",
       "      <th>AL645608.3</th>\n",
       "      <th>NOC2L</th>\n",
       "      <th>KLHL17</th>\n",
       "      <th>PLEKHN1</th>\n",
       "      <th>HES4</th>\n",
       "      <th>ISG15</th>\n",
       "      <th>AL645608.2</th>\n",
       "      <th>...</th>\n",
       "      <th>AC084398.2</th>\n",
       "      <th>OR4E1</th>\n",
       "      <th>PRR35</th>\n",
       "      <th>AC018553.2</th>\n",
       "      <th>AC005180.1</th>\n",
       "      <th>AC025048.1</th>\n",
       "      <th>AC015802.1</th>\n",
       "      <th>AP005137.2</th>\n",
       "      <th>AL121917.2</th>\n",
       "      <th>AC016590.3</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGCGTACA-1-sample42</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGAGGGAGTAA-1-sample42</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>1.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCAGCGATCC-1-sample42</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCATATGGTC-1-sample42</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACCTGCATTGAGCT-1-sample42</th>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 22091 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                             AL627309.1  AL669831.5  LINC00115  AL645608.3  \\\n",
       "AAACCTGAGGCGTACA-1-sample42         0.0         0.0        0.0         0.0   \n",
       "AAACCTGAGGGAGTAA-1-sample42         0.0         0.0        0.0         0.0   \n",
       "AAACCTGCAGCGATCC-1-sample42         0.0         0.0        0.0         0.0   \n",
       "AAACCTGCATATGGTC-1-sample42         0.0         0.0        0.0         0.0   \n",
       "AAACCTGCATTGAGCT-1-sample42         0.0         0.0        0.0         0.0   \n",
       "\n",
       "                             NOC2L  KLHL17  PLEKHN1  HES4  ISG15  AL645608.2  \\\n",
       "AAACCTGAGGCGTACA-1-sample42    0.0     0.0      0.0   0.0    1.0         0.0   \n",
       "AAACCTGAGGGAGTAA-1-sample42    0.0     0.0      0.0   0.0    1.0         0.0   \n",
       "AAACCTGCAGCGATCC-1-sample42    0.0     0.0      0.0   0.0    0.0         0.0   \n",
       "AAACCTGCATATGGTC-1-sample42    0.0     0.0      0.0   0.0    0.0         0.0   \n",
       "AAACCTGCATTGAGCT-1-sample42    0.0     0.0      0.0   0.0    0.0         0.0   \n",
       "\n",
       "                             ...  AC084398.2  OR4E1  PRR35  AC018553.2  \\\n",
       "AAACCTGAGGCGTACA-1-sample42  ...         0.0    0.0    0.0         0.0   \n",
       "AAACCTGAGGGAGTAA-1-sample42  ...         0.0    0.0    0.0         0.0   \n",
       "AAACCTGCAGCGATCC-1-sample42  ...         0.0    0.0    0.0         0.0   \n",
       "AAACCTGCATATGGTC-1-sample42  ...         0.0    0.0    0.0         0.0   \n",
       "AAACCTGCATTGAGCT-1-sample42  ...         0.0    0.0    0.0         0.0   \n",
       "\n",
       "                             AC005180.1  AC025048.1  AC015802.1  AP005137.2  \\\n",
       "AAACCTGAGGCGTACA-1-sample42         0.0         0.0         0.0         0.0   \n",
       "AAACCTGAGGGAGTAA-1-sample42         0.0         0.0         0.0         0.0   \n",
       "AAACCTGCAGCGATCC-1-sample42         0.0         0.0         0.0         0.0   \n",
       "AAACCTGCATATGGTC-1-sample42         0.0         0.0         0.0         0.0   \n",
       "AAACCTGCATTGAGCT-1-sample42         0.0         0.0         0.0         0.0   \n",
       "\n",
       "                             AL121917.2  AC016590.3  \n",
       "AAACCTGAGGCGTACA-1-sample42         0.0         0.0  \n",
       "AAACCTGAGGGAGTAA-1-sample42         0.0         0.0  \n",
       "AAACCTGCAGCGATCC-1-sample42         0.0         0.0  \n",
       "AAACCTGCATATGGTC-1-sample42         0.0         0.0  \n",
       "AAACCTGCATTGAGCT-1-sample42         0.0         0.0  \n",
       "\n",
       "[5 rows x 22091 columns]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "raw_counts = adata.layers[\"RNA_counts\"]\n",
    "\n",
    "df_counts = pd.DataFrame(\n",
    "    raw_counts[:10, ].toarray(),  # Convert to dense\n",
    "    index=adata.obs_names[:10],  # Cell barcodes\n",
    "    columns=adata.var_names  # Gene names\n",
    ")\n",
    "\n",
    "df_counts.head()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/albert.baichorov/.venv/lib/python3.10/site-packages/scvi/data/fields/_layer_field.py:115: UserWarning: Training will be faster when sparse matrix is formatted as CSR. It is safe to cast before model initialization.\n",
      "  _verify_and_correct_data_format(adata, self.attr_name, self.attr_key)\n"
     ]
    }
   ],
   "source": [
    "scvi.model.SCVI.setup_anndata(adata, batch_key=\"sampleName\", layer=\"RNA_counts\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "ename": "",
     "evalue": "",
     "output_type": "error",
     "traceback": [
      "\u001b[1;31mThe Kernel crashed while executing code in the current cell or a previous cell. \n",
      "\u001b[1;31mPlease review the code in the cell(s) to identify a possible cause of the failure. \n",
      "\u001b[1;31mClick <a href='https://aka.ms/vscodeJupyterKernelCrash'>here</a> for more info. \n",
      "\u001b[1;31mView Jupyter <a href='command:jupyter.viewOutput'>log</a> for further details."
     ]
    }
   ],
   "source": [
    "vae = scvi.model.SCVI(adata)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "vae.train()\n",
    "adata.obsm[\"X_scVI\"] = vae.get_latent_representation()\n",
    "adata.obsm[\"X_normalized_scVI\"] = vae.get_normalized_expression()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": ".venv",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
